granulate_utils/containers/client.py

```diff
@@ -0,0 +1,61 @@
+#
+# Copyright (c) Granulate. All rights reserved.
+# Licensed under the AGPL3 License. See LICENSE.md in the project root for license information.
+#
+
+from typing import List, Optional
+
+from granulate_utils.containers.container import Container
+from granulate_utils.containers.cri import CriClient
+from granulate_utils.containers.docker import DockerClient
+from granulate_utils.exceptions import NoContainerRuntimesError
+
+
+class ContainersClient:
+    """
+    Wraps DockerClient and CriClient to provide a unified view of all containers
+    running on a system.
+    Docker is going away in k8s (https://kubernetes.io/blog/2020/12/02/dont-panic-kubernetes-and-docker/)
+    so this is our way to collect all containers in either case, whether Docker is used on this system
+    or not.
+    """
+
+    def __init__(self) -> None:
+        try:
+            self._docker_client: Optional[DockerClient] = DockerClient()
+        except Exception:
+            self._docker_client = None
+
+        try:
+            self._cri_client: Optional[CriClient] = CriClient()
+        except Exception:
+            self._cri_client = None
+
+        if self._docker_client is None and self._cri_client is None:
+            raise NoContainerRuntimesError()
+
+    def list_containers(self, all_info: bool = False) -> List[Container]:
+        """
+        Lists all containers running on this machine via DockerClient and CriClient.
+        :param all_info: Collect more verbose information. Currently, this ensures that the pid field of each
+                         Container object is filled in.
+        """
+        docker_containers = self._docker_client.list_containers(all_info) if self._docker_client is not None else []
+        cri_containers = self._cri_client.list_containers(all_info) if self._cri_client is not None else []
+
+        # start with all Docker containers
+        containers = docker_containers.copy()
+        # then add CRI containers that are not already listed
```

Comment:
There is, will document

Reply:
> CRI won't list containers controlled by Docker

Did you mean "Docker won't list all containers controlled by containerd"? Because otherwise, I don't understand why would containerd not list docker started containers

Reply:
CRI is a plugin of containerd, and it will list containerd containers that were started by CRI.
Docker also uses containerd but not via CRI, so CRI is invisible to Docker containers, and Docker is invisible to CRI containers (in containerd level, they run in different namespaces - CRI uses `k8s.io` and Docker uses `moby`).

Also - we might use cri-o in which case it doesn't use containerd (AFAIK) and if there Docker & containerd, they are entirely unrelated. Though I haven't seen such cases yet.

Will clarify a bit more.

Reply:
Quite unexpected, can you add: "[..] by use of different namespaces in containerd - https://github.com/containerd/containerd/blob/main/docs/namespaces.md"

Reply:
Hmm, I assume it's the namespace, but I haven't checked CRI's code. Fact is, it won't reply with containers created by Docker, and it makes sense to me. I can note it with less certainty.

Reply:
You're right,  
In containerd CRI plugin - [here](https://github.com/containerd/containerd/blob/main/pkg/cri/cri.go#L92) and [here](https://github.com/containerd/containerd/blob/main/pkg/cri/constants/constants.go#L21)

And for docker, I ran `sudo ctr namespaces list` and got:
![image](https://user-images.githubusercontent.com/73543893/154854831-27be41af-c1d9-42d4-9522-356080c34359.png)

So, I think it's safe to say they use namespaces for that